### PR TITLE
GH-127010: Don't lazily track and untrack dicts

### PR DESCRIPTION
**Pull request by markshannon**
Remove the code to lazily track and untrack dictionaries.

[Benchmarks](https://github.com/faster-cpython/benchmarking-public/tree/main/results/bm-20241119-3.14.0a1%2B-498979f) show a ~1% speedup.

* Issue: gh-127010

----
📚 Documentation preview 📚: https://cpython-previews--127027.org.readthedocs.build/

### CODE REVIEW

File: InternalDocs/garbage_collector.md [hunk 1]
```diff
@@ -581,8 +575,6 @@ tracking status of the object.
       >>> gc.is_tracked({})
       False
```
Comment (methane) [inline InternalDocs/garbage_collector.md:573]: Is empty dict still untracked?

Comment (markshannon) [reply InternalDocs/garbage_collector.md:573]: No it is tracked. I'll update the docs.

**Comment by markshannon** [comment]
The windows buildbot failures are unrelated to this PR.

File: InternalDocs/garbage_collector.md [hunk 2]
```diff
@@ -548,8 +548,8 @@ a container:
 As a general rule, instances of atomic types aren't tracked and instances of
 non-atomic types (containers, user-defined objects...) are.  However, some
 type-specific optimizations can be present in order to suppress the garbage
-collector footprint of simple instances. Some examples of native types that
-benefit from delayed tracking:
+collector footprint of simple instances. Historically, both dictionaries and
+tuples were untracked during garbage collection. Now it is only tuples:
 
 - Tuples containing only immutable objects (integers, strings etc,
```
Comment (iritkatriel) [inline InternalDocs/garbage_collector.md]: This is now a single-item bullet list.

File: Objects/dictobject.c [hunk 3]
```diff
@@ -7261,12 +7194,6 @@ _PyDict_DetachFromObject(PyDictObject *mp, PyObject *obj)
     PyDictValues *values = copy_values(mp->ma_values);
 
     if (values == NULL) {
-        /* Out of memory. Clear the dict */
-        PyInterpreterState *interp = _PyInterpreterState_GET();
-        PyDictKeysObject *oldkeys = mp->ma_keys;
-        set_keys(mp, Py_EMPTY_KEYS);
-        dictkeys_decref(interp, oldkeys, IS_DICT_SHARED(mp));
-        STORE_USED(mp, 0);
         PyErr_NoMemory();
```
Comment (iritkatriel) [inline Objects/dictobject.c:7197]: Wouldn't `copy_values` already have set the memory error?

Comment (markshannon) [reply Objects/dictobject.c:7197]: No. https://github.com/python/cpython/blob/main/Objects/dictobject.c#L3974

**Review by iritkatriel — APPROVED** [review]

**Comment by markshannon** [comment]
The `aarch64-unknown-linux-gnu/gcc (Debug)` failures seem unrelated to this PR (probably timeout related) and there are no leaks.

**Pull request merged**